### PR TITLE
Add additional password hash algorithms

### PR DESCRIPTION
**Pull request by EpicCoder**
Added addional (bcrypt, scrypt, argon) password hashing algorithms to gitea. It can be set in app.ini 

### Dep
Added new dependency for the algorithms and updated 'golang.org/x/crypto'

### Test
Updatet tests for the new algorithms 

###  Doc
Added description and available algorithms for documentation and add example in app.ini.sample

### CODE REVIEW

**Review by lafriks — APPROVED** [review]

File: custom/conf/app.ini.sample
Comment (lunny) [inline custom/conf/app.ini.sample]: use `=` not `:`

File: models/user.go
Comment (lunny) [inline models/user.go]: The method should only do validate but not update password.

**Review by lunny — CHANGES_REQUESTED** [review]

File: custom/conf/app.ini.sample
Comment (EpicCoder) [reply custom/conf/app.ini.sample]: oh, i correct this

File: models/user.go
Comment (EpicCoder) [reply models/user.go]: I think its useful to automatically update the password hash when the hash algo was changed without user intervention and only if the passsword is correct. Is there a better way to do this?

Comment (lafriks) [reply models/user.go]: I agree that it is ok to update to new hashing password if server setting has changed

**Comment by lafriks** [comment]
Please resolve conflicts

File: models/user.go
Comment (adelowo) [inline models/user.go]:
```suggestion
// HashPassword hashes a password using the algorithm defined in the config value of PASSWORD_HASH_ALGO.

```

**Review by adelowo — APPROVED** [review]

File: models/user.go
Comment (kolaente) [inline models/user.go]: Please use constants for the password types.

File: models/user.go
Comment (kolaente) [inline models/user.go]: Please correct me if I'm wrong, but doesn't this just always use `bcrypt`?

**Comment by kolaente** [comment]
Please resolve conflicts.

File: models/user.go
Comment (EpicCoder) [reply models/user.go]: Yes it does, but it think its not a problem because it return only true if the stored hash is a Bcrypt hash and match the entered password.

**Comment by EpicCoder** [comment]
Resolved conflicts.

**Comment by lafriks** [comment]
@lunny please review

File: models/user.go
Comment (lunny) [reply models/user.go]: Sorry, I can't agree with you. If you want to to update database on a function name has `validate` prefix, please rename it.

Comment (EpicCoder) [reply models/user.go]: @lunny Should i rename the function from "ValidatePassword" to "UpdatePasswordAndValidate"?

Comment (lunny) [reply models/user.go]: As my personal point, I would like to keep the name `ValidatePassword` and change the logic where invoke it.

**Comment by EpicCoder** [comment]
Moved the update logic out of ValidatePassword @lunny please review

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@EpicCoder Thanks! 
Maintainers notice: Please keep this not merge until branch v1.9 created since it has been moved to v1.10.

**Pull request merged**
